src/wallet/scriptpubkeyman.cpp

```diff
@@ -1717,42 +1717,190 @@ std::unordered_set<CScript, SaltedSipHasher> LegacyDataSPKM::GetScriptPubKeys()
+
+    // mapScripts is iterated to compute all additional spendable output scripts that utilize the contained scripts
+    // as redeemScripts and witnessScripts.
+    //
+    // mapScripts contains redeemScripts and witnessScripts. It may also contain output scripts which,
+    // in addition to being treated as output scripts, are also treated as redeemScripts and witnessScripts.
+    // All scripts in mapScripts are treated as redeemScripts, unless that script is also a P2SH.
+    // A script is only treated as a witnessScript if its corresponding P2WSH output script is in the map.
+    for (const auto& [_, script] : mapScripts) {
+        std::vector<std::vector<unsigned char>> solutions;
+        TxoutType type = Solver(script, solutions);
+        switch (type) {
+        // We don't care about these types because they are not spendable
+        case TxoutType::NONSTANDARD:
+        case TxoutType::NULL_DATA:
+        case TxoutType::WITNESS_UNKNOWN:
+        case TxoutType::ANCHOR:
+        // P2TR are not spendable as the legacy wallet never supported them.
+        case TxoutType::WITNESS_V1_TAPROOT:
+        // These are only spendable if the witness script is also spendable as a scriptPubKey
+        // We will check these later after "spks" has been updated with the computed output scripts from mapScripts.
+        case TxoutType::WITNESS_V0_SCRIPTHASH:
+        // For P2SH to be spendable, we need to have the redeemScript, and if we have it, it will be handled when
+        // this loop gets to it. A P2SH script itself cannot be nested in anything, so we can skip them.
+        case TxoutType::SCRIPTHASH:
+            break;
+        // Any P2PK or P2PKH scripts found in mapScripts can be spent as P2SH-P2PK or P2SH-P2PKH respectively,
+        // if we have the private key.
+        // Since all private keys were iterated earlier and their corresponding P2PK and P2PKH scripts inserted
+        // to "spks", we can simply check whether this P2PK or P2PKH script is in "spks" to determine spendability.
+        case TxoutType::PUBKEY:
+        case TxoutType::PUBKEYHASH:
+            if (spks.count(script) > 0) {
                 spks.insert(GetScriptForDestination(ScriptHash(script)));
             }
-            // For segwit scripts, we only consider them spendable if we have the segwit spk
-            int wit_ver = -1;
-            std::vector<unsigned char> witprog;
-            if (script.IsWitnessProgram(wit_ver, witprog) && wit_ver == 0) {
+            break;
+        // P2WPKH scripts are only spendable if we have the private key.
+        case TxoutType::WITNESS_V0_KEYHASH:
+        {
+            CKeyID key_id{uint160(solutions[0])};
+            CPubKey pubkey;
+            if (GetPubKey(key_id, pubkey) && pubkey.IsCompressed() && HaveKey(key_id)) {
                 spks.insert(script);
+                // Also insert P2SH-P2WPKH output script
+                spks.insert(GetScriptForDestination(ScriptHash(script)));
```

Comment:
Why are we not generating the relevant spks for all keys directly in the initial loop? It would just mean that we would now perhaps recognize and be able to spend more output script types than before, so they should simply not appear in our history, but it might make the code simpler?

Reply:
Done